src/miner.cpp

```diff
@@ -206,68 +203,73 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
                     break;
                 }
                 continue;
             }
 
-            CAmount nTxFees = iter->GetFee();
             // Added
-            pblock->vtx.push_back(tx);
-            pblocktemplate->vTxFees.push_back(nTxFees);
-            pblocktemplate->vTxSigOps.push_back(nTxSigOps);
+            block.AddTx(std::move(tx));
             nBlockSize += nTxSize;
             ++nBlockTx;
             nBlockSigOps += nTxSigOps;
-            nFees += nTxFees;
+            nFees += tx.GetFee();
 
             inBlock.insert(iter);
         }
 
-
-        if (IsFourthHFActive(nMedianTimePast)) {
-            // If magnetic anomaly is enabled, we make sure transaction are
-            // lexically ordered.
-            std::sort(std::begin(pblock->vtx) + 1, std::end(pblock->vtx),
-                    [](const CTransaction& a, const CTransaction& b) -> bool {
-                        return a.GetHash() < b.GetHash();
-                    });
-        }
-
         nLastBlockTx = nBlockTx;
         nLastBlockSize = nBlockSize;
         LogPrintf("CreateNewBlock(): total size %u txs: %u fees: %ld sigops %d\n", nBlockSize, nBlockTx, nFees, nBlockSigOps);
 
         // Compute final coinbase transaction.
         txNew.vout[0].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus());
         txNew.vin[0].scriptSig = CScript() << nHeight << BIP100Str(hardLimit) << OP_0;
-        BloatCoinbaseSize(txNew);
-        pblock->vtx[0] = txNew;
-        pblocktemplate->vTxFees[0] = -nFees;
+        miner::BloatCoinbaseSize(txNew);
+        CTransaction coinbaseTx(txNew);
+        miner::BuilderEntry coinbase(&coinbaseTx, GetLegacySigOpCount(txNew, STANDARD_CHECKDATASIG_VERIFY_FLAGS), 0);
+        block.SetCoinbase(std::move(coinbase));
 
         // Fill in header
-        pblock->hashPrevBlock  = pindexPrev->GetBlockHash();
-        UpdateTime(pblock, Params().GetConsensus(), pindexPrev);
-        pblock->nBits          = GetNextWorkRequired(pindexPrev, pblock, Params().GetConsensus());
-        pblock->nNonce         = 0;
-        pblocktemplate->vTxSigOps[0] = GetLegacySigOpCount(pblock->vtx[0],
-                                                           STANDARD_CHECKDATASIG_VERIFY_FLAGS);
-
-        CValidationState state;
-        if (!TestBlockValidity(state, *pblock, pindexPrev, false, false)) {
-            throw std::runtime_error(strprintf("%s: TestBlockValidity failed: %s", __func__, FormatStateMessage(state)));
+        block.SetHashPrevBlock(pindexPrev->GetBlockHash());
+        UpdateTime(block, Params().GetConsensus(), pindexPrev);
+        block.SetBits(GetNextWorkRequired(pindexPrev, block.GetTime(), Params().GetConsensus()));
+
+        if (block.NeedsGBTMetadata()) {
+            block.SetBlockMinTime(nMedianTimePast + 1);
+            block.SetBlockHeight(nHeight);
+            block.SetBlockSizeLimit(hardLimit);
+            block.SetBlockSigopLimit(MaxBlockSigops(nBlockSize));
```

Comment:
This is different from before, it used to be `MaxBlockSigops(hadLimit)`. I think it is more accurate to be on the safe side here.

Reply:
Since sigoplimit grows with block size, with `hardLimit` he'd have to stuff the block completely full to be sure he's not exceeding the limit when adding transactions from another source, right?

Reply:
You're right of course.  How quickly I forgot the decision to validate using the actual block size.